strawberryfields/circuitspecs/xstrict.py

```diff
@@ -0,0 +1,183 @@
+        n_modes = len(registers)
+
+        # Number of modes must be even
+        if n_modes % 2 != 0:
+            raise CircuitError("The X series only supports programs with an even number of modes.")
+        half_n_modes = n_modes // 2
+        # Call the GBS compiler to do basic measurement validation.
+        # The GBS compiler also merges multiple measurement commands
+        # into a single MeasureFock command at the end of the circuit.
+        seq = GBSSpecs().compile(seq, registers)
+
+        # ensure that all modes are measured
+        if len(seq[-1].reg) != n_modes:
+            raise CircuitError("All modes must be measured.")
+
+        # Check circuit begins with two-mode squeezers
+        # --------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+        # If there are no two-mode squeezers add squeezers at the beginning with squeezing param equal to zero.
+        if B == []:
+            initS2 = [
+                Command(ops.S2gate(0, 0), [registers[i], registers[i + half_n_modes]])
+                for i in range(half_n_modes)
+            ]
+            seq = initS2 + seq
+            A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+
+        if A != []:
+            raise CircuitError("There can be no operations before the S2gates.")
+
+        regrefs = set()
+
+        if B:
+            # get set of circuit registers as a tuple for each S2gate
+            regrefs = {(cmd.reg[0].ind, cmd.reg[1].ind) for cmd in B}
+
+        # the set of allowed mode-tuples the S2gates must have
+        allowed_modes = set(zip(range(0, half_n_modes), range(half_n_modes, n_modes)))
+
+        if not regrefs.issubset(allowed_modes):
+            raise CircuitError("S2gates do not appear on the correct modes.")
+
+        # determine which modes do not have input S2gates specified
+        missing = allowed_modes - regrefs
+
+        for i, j in missing:
+            # insert S2gates with 0 squeezing
+            B.insert(0, Command(ops.S2gate(0, 0), [registers[i], registers[j]]))
+
+        sqs = [cmd.op.p[0] for cmd in B]
+
+        # ensure provided S2gates all have the allowed squeezing values
+        if not all(s in self.allowed_sq_ranges for s in sqs):
+            wrong_sq_values = [np.round(s, 4) for s in sqs if s not in self.allowed_sq_ranges]
+            raise CircuitError(
+                "Incorrect squeezing value(s) r={}. Allowed squeezing "
+                "value(s) are {}.".format(wrong_sq_values, self.allowed_sq_ranges)
+            )
+        # This could in principle be changed
+        phases = [cmd.op.p[1] for cmd in B]
+        if not np.allclose(phases, 0):
+            raise CircuitError(
+                "Incorrect phase value(s) phi={}. Allowed squeezing "
+                "value(s) are 0.0.".format(phases)
+            )
+
+        meas_seq = [C[-1]]
+        seq = GaussianUnitary().compile(C[:-1], registers)
```

Comment:
Nice 🙂

Reply:
Can you remove the `[:-1]` simply by adding `MeasureFock` to the primitives of the `GaussianCompiler`?